### PR TITLE
Customizing whether passengers are kicked out when an aircraft fires

### PR DESCRIPTION
**Pull request by ststl-s**
- In yuri's revenge, if aircraft has passengers, aircraft force kick out it's passengers when firing but not fire weapons.
- Now you can decide aircraft kick out it's passengers or not when firing.

In `rulesmd.ini`
```ini
[SOMEWEAPON]
KickOutPassenger=true  ; boolean 
```

### CODE REVIEW

**Comment by itsyuribri** [comment]
Weapon code = 

[LaserGun]
Damage=120
ROF=20
Range=8
Projectile=InvisibleLaser
Speed=100
Report=Laser
Warhead=LaserGunWH
Bright=yes
IsHouseColor=false
LaserInnerColor=200,0,0
LaserOuterColor=200,0,0
LaserDuration=10
IsLaser=true
Burst=2
Anim=LASMUZ
Ammo=1
KickOutPassenger=false



[snapshot-20230911-141542.zip](https://github.com/Phobos-developers/Phobos/files/12579225/snapshot-20230911-141542.zip)

**Comment by itsyuribri** [comment]
https://drive.google.com/file/d/150L4gLWvvgZ54WaRU8wgJyBIxXi0T3Oq/view?usp=drive_link

**Comment by ststl-s** [comment]
I con't download that, please send email to `ststl-s@outlook.com`.

> https://drive.google.com/file/d/150L4gLWvvgZ54WaRU8wgJyBIxXi0T3Oq/view?usp=drive_link

File: docs/Fixed-or-Improved-Logics.md [hunk 1]
```diff
@@ -859,6 +859,17 @@ ShakeIsLocal=false  ; boolean
 
 ## Weapons
 
+### Customizing whether passengers are kicked out when an aircraft fires
+
+- In yuri's revenge, if aircraft has passengers, aircraft force kick out it's passengers when firing but not fire weapons.
+- Now you can decide aircraft kick out it's passengers or not when firing.
```
Comment (MortonPL) [inline docs/Fixed-or-Improved-Logics.md:865]:
```suggestion
- You can now customize whether aircraft will forcefully eject passengers (vanilla behavior) or fire its weapon when attempting to fire.
```

File: docs/Fixed-or-Improved-Logics.md [hunk 2]
```diff
@@ -859,6 +859,17 @@ ShakeIsLocal=false  ; boolean
...
+
+In `rulesmd.ini`
+```ini
+[SOMEWEAPON]
+KickOutPassenger=true  ; boolean
```
Comment (MortonPL) [inline docs/Fixed-or-Improved-Logics.md:870]:
```suggestion
KickOutPassengers=true  ; boolean
```

File: src/Ext/WeaponType/Body.h [hunk 3]
```diff
@@ -60,6 +61,7 @@ class WeaponTypeExt
 			, OmniFire_TurnToTarget { false }
 			, ExtraWarheads {}
 			, ExtraWarheads_DamageOverrides {}
+			, KickOutPassenger { true }
```
Comment (MortonPL) [inline src/Ext/WeaponType/Body.h:64]:
```suggestion
			, KickOutPassengers { true }
```

File: src/Ext/WeaponType/Body.h [hunk 4]
```diff
@@ -39,6 +39,7 @@ class WeaponTypeExt
 		Valueable<bool> OmniFire_TurnToTarget;
 		ValueableVector<WarheadTypeClass*> ExtraWarheads;
 		ValueableVector<int> ExtraWarheads_DamageOverrides;
+		Valueable<bool> KickOutPassenger;
```
Comment (MortonPL) [inline src/Ext/WeaponType/Body.h:42]:
```suggestion
		Valueable<bool> KickOutPassengers;
```

File: src/Ext/WeaponType/Body.cpp [hunk 5]
```diff
@@ -84,6 +85,7 @@ void WeaponTypeExt::ExtData::Serialize(T& Stm)
 		.Process(this->OmniFire_TurnToTarget)
 		.Process(this->ExtraWarheads)
 		.Process(this->ExtraWarheads_DamageOverrides)
+		.Process(this->KickOutPassenger)
```
Comment (MortonPL) [inline src/Ext/WeaponType/Body.cpp:88]:
```suggestion
		.Process(this->KickOutPassengers)
```

File: src/Ext/WeaponType/Body.cpp [hunk 6]
```diff
@@ -59,6 +59,7 @@ void WeaponTypeExt::ExtData::LoadFromINIFile(CCINIClass* const pINI)
 	this->OmniFire_TurnToTarget.Read(exINI, pSection, "OmniFire.TurnToTarget");
 	this->ExtraWarheads.Read(exINI, pSection, "ExtraWarheads");
 	this->ExtraWarheads_DamageOverrides.Read(exINI, pSection, "ExtraWarheads.DamageOverrides");
+	this->KickOutPassenger.Read(exINI, pSection, "KickOutPassenger");
```
Comment (MortonPL) [inline src/Ext/WeaponType/Body.cpp:62]:
```suggestion
	this->KickOutPassenger.Read(exINI, pSection, "KickOutPassengers");
```

File: src/Ext/Aircraft/Hooks.cpp [hunk 7]
```diff
@@ -107,3 +110,27 @@ DEFINE_HOOK(0x414C0B, AircraftClass_ChronoSparkleDelay, 0x5)
...
+		const WeaponTypeClass* pWeapon = pThis->GetWeapon(weaponIdx)->WeaponType;
+
+		if (const auto pWeaponExt = WeaponTypeExt::ExtMap.Find(pWeapon))
+		{
+			if (!pWeaponExt->KickOutPassenger)
+				return Fire;
+		}
+	}
+
+	return KickOutPassenger;
```
Comment (MortonPL) [inline src/Ext/Aircraft/Hooks.cpp:135]:
```suggestion
	if (!pThis->Passengers.GetFirstPassenger())
		return Fire;

	const auto pWeapon = pThis->GetWeapon(weaponIdx);
	if (!TechnoExt::IsActive(pThis) || !pWeapon)
		return KickOutPassengers;

	const auto pWeaponTypeExt = WeaponTypeExt::ExtMap.Find(pWeapon->WeaponType);
	if (!pWeaponTypeExt)
		return KickOutPassengers;

	return pWeaponExt->KickOutPassengers ? KickOutPassengers : Fire;
```

**Review by MortonPL — CHANGES_REQUESTED** [review]

**Comment by ststl-s** [comment]
Please test the latest version. If a fatal error occurs again, send me an email. Thank you. @itsyuribri

**Comment by itsyuribri** [comment]
> Please test the latest version. If a fatal error occurs again, send me an email. Thank you. @itsyuribri [Phobos-pr1151-86ad0acc.zip](https://github.com/Phobos-developers/Phobos/files/12605452/Phobos-pr1151-86ad0acc.zip)

Nope, still crashes

[LaserGun]
Damage=120
ROF=20
Range=8
Projectile=InvisibleStorm
Speed=100
Report=Laser
Warhead=LaserGunWH
Bright=yes
IsHouseColor=false
LaserInnerColor=200,0,0
LaserOuterColor=200,0,0
LaserDuration=10
IsLaser=true
Burst=2
Anim=LASMUZ
Ammo=1
KickOutPassenger=false

dump link 
https://drive.google.com/file/d/1Id0jmLCNThpzZhy8gLHqIRybYkEZV0Nj/view?usp=drive_link

**Comment by ststl-s** [comment]
> > Please test the latest version. If a fatal error occurs again, send me an email. Thank you. @itsyuribri [Phobos-pr1151-86ad0acc.zip](https://github.com/Phobos-developers/Phobos/files/12605452/Phobos-pr1151-86ad0acc.zip)
> 
> Nope, still crashes
> 
> [LaserGun] Damage=120 ROF=20 Range=8 Projectile=InvisibleStorm Speed=100 Report=Laser Warhead=LaserGunWH Bright=yes IsHouseColor=false LaserInnerColor=200,0,0 LaserOuterColor=200,0,0 LaserDuration=10 IsLaser=true Burst=2 Anim=LASMUZ Ammo=1 KickOutPassenger=false
> 
> dump link https://drive.google.com/file/d/1Id0jmLCNThpzZhy8gLHqIRybYkEZV0Nj/view?usp=drive_link

**Comment by ststl-s** [comment]
There is no dump file in this link, and it seems like the one you tested on September 11th before. @itsyuribri

**Comment by itsyuribri** [comment]
> There is no dump file in this link, and it seems like the one you tested on September 11th before. @itsyuribri

It still crashes, and the same weapon code is used, I downloaded the latest one that you sent me. 

https://drive.google.com/file/d/16UonpqpE6sDwMrW6eYDEux4wxamg0IqZ/view?usp=drive_link

**Comment by itsyuribri** [comment]
It still crashes, and the same weapon code is used, I downloaded the latest
one that you sent me.

https://drive.google.com/file/d/16UonpqpE6sDwMrW6eYDEux4wxamg0IqZ/view?usp=drive_link

**Comment by ststl-s** [comment]
This dump file can't find reason, but other 3 testers and I didn't crash with YR or Mental Omega or other mods.

**Comment by Starkku** [comment]
Rebased on develop to fix merge conflicts, amended the hook to remove some unnecessary checks / clean it up. Good to go now so merging.

**Pull request merged**